File: vane-core/src/main/java/org/oddlama/vane/core/config/recipes/ShapelessRecipeDefinition.java
package org.oddlama.vane.core.config.recipes;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.Tag;
import org.bukkit.inventory.Recipe;
import org.bukkit.inventory.ShapelessRecipe;
import org.oddlama.vane.util.ItemUtil;

public class ShapelessRecipeDefinition extends RecipeDefinition {

    private List<String> ingredients = new ArrayList<>();
    private String result = null;

    public ShapelessRecipeDefinition(String name) {
        super(name);
    }

    public ShapelessRecipeDefinition addIngredient(String ingredient) {
        this.ingredients.add(ingredient);
        return this;
    }

    public ShapelessRecipeDefinition addIngredient(final Tag<?> tag) {
        return addIngredient("#" + tag.key());
    }

    public ShapelessRecipeDefinition addIngredient(Material material) {
        return addIngredient(material.key().toString());
    }

    public ShapelessRecipeDefinition result(String result) {
        this.result = result;
        return this;
    }

    @Override
    public Object toDict() {
        final HashMap<String, Object> dict = new HashMap<>();
        dict.put("Type", "shapeless");
        dict.put("Ingredients", this.ingredients);
        dict.put("Result", this.result);
        return dict;
    }

    @Override
    public RecipeDefinition fromDict(Object dict) {
        if (!(dict instanceof Map<?, ?>)) {
            throw new IllegalArgumentException(
                "Invalid shapeless recipe dictionary: Argument must be a Map<String, Object>!"
            );
        }
        final var dictMap = (Map<?, ?>) dict;
        final Object ingredientsObj = dictMap.containsKey("Ingredients") ? dictMap.get("Ingredients") : dictMap.get("ingredients");
        if (ingredientsObj instanceof List<?> ingredients) {
            this.ingredients = ingredients.stream().map(i -> (String) i).toList();
        } else {
            throw new IllegalArgumentException(
                "Invalid shapeless recipe dictionary: ingredients must be a list of strings"
            );
        }

        final Object resultObj = dictMap.containsKey("Result") ? dictMap.get("Result") : dictMap.get("result");
        if (resultObj instanceof String result) {
            this.result = result;
        } else {
            throw new IllegalArgumentException("Invalid shapeless recipe dictionary: result must be a string");
        }

        return this;
    }

    @Override
    public Recipe toRecipe(NamespacedKey baseKey) {
        final var recipe = new ShapelessRecipe(key(baseKey), ItemUtil.itemstackFromString(this.result).getLeft());
        this.ingredients.forEach(i -> recipe.addIngredient(RecipeDefinition.recipeChoice(i)));
        return recipe;
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/recipes/ShapelessRecipeDefinition.kt
package org.oddlama.vane.core.config.recipes

import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.Tag
import org.bukkit.inventory.Recipe
import org.bukkit.inventory.ShapelessRecipe
import org.oddlama.vane.util.ItemUtil

/**
 * Recipe definition for shapeless crafting recipes.
 *
 * @param name logical recipe name.
 */
class ShapelessRecipeDefinition(name: String?) : RecipeDefinition(name) {
    /** Ingredient definition strings. */
    private var ingredients: MutableList<String> = ArrayList()
    /** Result item definition string. */
    private var result: String? = null

    /** Adds an ingredient definition string. */
    fun addIngredient(ingredient: String): ShapelessRecipeDefinition {
        this.ingredients.add(ingredient)
        return this
    }

    /** Adds an ingredient definition from a Bukkit tag. */
    fun addIngredient(tag: Tag<*>): ShapelessRecipeDefinition {
        return addIngredient("#" + tag.key())
    }

    /** Adds an ingredient definition from a material key. */
    fun addIngredient(material: Material): ShapelessRecipeDefinition {
        return addIngredient(material.key().toString())
    }

    /** Sets the result item definition string. */
    fun result(result: String?): ShapelessRecipeDefinition {
        this.result = result
        return this
    }

    /** Serializes this recipe definition to dictionary form. */
    override fun toDict(): Any {
        val dict = mutableMapOf<String?, Any?>()
        dict["Type"] = "shapeless"
        dict["Ingredients"] = this.ingredients
        dict["Result"] = this.result
        return dict
    }

    /** Loads this recipe definition from dictionary form. */
    override fun fromDict(dict: Any?): RecipeDefinition {
        require(dict is MutableMap<*, *>) { "Invalid shapeless recipe dictionary: Argument must be a Map<String, Object>!" }
        val ingredientsObj =
            if (dict.containsKey("Ingredients")) dict["Ingredients"] else dict["ingredients"]
        if (ingredientsObj is MutableList<*>) {
            this.ingredients = ingredientsObj.filterIsInstance<String>().toMutableList()
        } else {
            throw IllegalArgumentException(
                "Invalid shapeless recipe dictionary: ingredients must be a list of strings"
            )
        }

        val resultObj = if (dict.containsKey("Result")) dict["Result"] else dict["result"]
        if (resultObj is String) {
            this.result = resultObj
        } else {
            throw IllegalArgumentException("Invalid shapeless recipe dictionary: result must be a string")
        }

        return this
    }

    /** Converts this definition into a Bukkit [ShapelessRecipe]. */
    override fun toRecipe(baseKey: NamespacedKey?): Recipe {
        val bk = baseKey ?: throw IllegalArgumentException("baseKey cannot be null")
        val recipe = ShapelessRecipe(key(bk), ItemUtil.itemstackFromString(this.result!!).getLeft()!!)
        this.ingredients.forEach { i -> recipe.addIngredient(recipeChoice(i)) }
        return recipe
    }
}
